File: quest/src/store/rewardEngine.js
const REWARD_CHANCE = 0.8;

export function rollReward(rewards) {
  const active = rewards.filter(r => r.active);
  if (active.length === 0) return { granted: false, reward: null };

  if (Math.random() > REWARD_CHANCE) {
    return { granted: false, reward: null };
  }

  const picked = active[Math.floor(Math.random() * active.length)];
  return { granted: true, reward: picked };
}

File: quest-ios/Sources/Models/RewardEngine.swift
import Foundation

enum RewardEngine {
    static func rollReward(_ rewards: [Reward]) -> (granted: Bool, reward: Reward?) {
        let active = rewards.filter(\.active)
        guard !active.isEmpty else { return (false, nil) }
        if Double.random(in: 0...1) > 0.8 { return (false, nil) }
        return (true, active.randomElement())
    }
}
